sql: add LTREE key encoding

paulniziolek opened the pull request:
#### sql: add LTREE keyside encoding 

This adds basic keyside encoding for LTREE, which is encoded similarly to a string array: ltrees have their own terminator and their labels are encoded via escaped bytes.

Informs: cockroachdb#44657
Epic: CRDB-148

Release note: None

paulniziolek commented on pkg/util/encoding/encoding.go:
```diff
@@ -136,10 +136,18 @@ const (
 	jsonKeyDescendingTerminator byte = 0xFF
 
+	// Terminators for LTREE Key encoding.
+	ltreeKeyTerminator           byte = 0x0F
+	ltreeKeyDescendingTerminator byte = 0xF0
```
To clarify, currently, an example ltree key encoding for 'A.B' looks like:
```
[ ltree marker ] 
    [ bytes marker ] 'A' [ bytes escape = 0x00 ] [ bytes terminator = 0x01 ]
    [ bytes marker ] 'B' [ bytes escape = 0x00 ] [ bytes terminator = 0x01 ]
[ ltree terminator = 0x0F ]
```

The alternative I proposed, if it is preferred for the LTREE terminator to be more consistent (aka have it be 0x00) would probably look like:
```
[ ltree marker ] 
    [ bytes marker ] 'A' [ bytes escape = 0x01 ] [ bytes terminator = 0x02 ]
    [ bytes marker ] 'B' [ bytes escape = 0x01 ] [ bytes terminator = 0x02 ]
[ ltree terminator = 0x00 ]
```
where the bytes have an adjusted escape + terminator to account for the reserved `0x00` for the ltree terminator. 

The reason why (i think) we shouldn't use `0x00` for both the ltree and bytes within the ltree is because this would make `PeekLength` slightly more complicated (and slower) since it would need to index through all of the byte terminators just to retrieve the length of the LTREE. 

Otherwise, I think the way key encoding is currently done seems simple as we don't need to define some new set of escapes/terminators for each LTREE label.

paulniziolek replied:
To reduce overhead per-label, we can omit the bytes encoding method and just encode the labels as-is with just a single label-specific terminator. This would reduce the overhead to 1 byte per label for encoding. 
<!-- Reviewable comment -OXUlzul9ekcK3dH2DKT:b-2sd3m5 -->
<!-- Sent from Reviewable.io -->

paulniziolek commented:
Currently, this index doesn't align perfectly because our `Prev` datum method implementation for LTREE is wrong. 

For example, whenever given some LTREE value `A.A`, the immediate correct "previous" ltree value would be...
`A.999999999999.zzzzzzzzzz.zzzzzzzzzz.zzzzzzzzzz...`, where this is just a short hand version, fully expanded this would have 1k max characters for each of the shorthand-written labels, and up to 65k labels...

yuzefovich reviewed (commented):
Nice! I think your solution works, but we could reduce the overhead (currently at 3 bytes per label) further - I left more context in a comment. It would also be nice to add some testing for this - like generate a couple of random LTREEs, then ensure the comparison between them is the same when comparing their key-encoding, and also check that we can decode the key-encoding into the equivalent to the original one.

2nd commit LGTM, but let's extract it into a separate PR so that it doesn't interfere with the main change.

> Reviewed 9 of 9 files at r1, 3 of 3 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/151597#-:-OXQabka6PT_5S2nuD6l:b-vhi8c)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @paulniziolek)

___
*[`pkg/sql/rowenc/keyside/encode.go` line 187 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/151597#-OXQUm43-hB70rfvMBwo:-OXQUm446XQZ536-ZIaB:bymh4sz) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed9b0e5a986f8de07a05b485f57e9829a41813de/pkg/sql/rowenc/keyside/encode.go#L187)):*
> ```Go
> 	if buildutil.CrdbTestBuild {
> 		if _, isLTree := val.(*tree.DLTree); isLTree {
> 			// TODO(paulniziolek): remove this exception once key encoding is
> ```

nit: we need to update `mustUseValueEncodingForFingerprinting`.
___
*[`pkg/util/encoding/encoding.go` line 139 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/151597#-OXQVs0rAoxo-ZpQThx5:-OXQVs0rAoxo-ZpQThx6:b-57hw0c) ([raw file](https://github.com/cockroachdb/cockroach/blob/fedea9cd7194c007941f21979b12702c034c5c3a/pkg/util/encoding/encoding.go#L139)):*
> ```Go
> 	jsonObjectKeyDescendingMarker     = jsonArrayKeyDescendingMarker - 1
> 
> 	// LTREE key encoding markers
> ```

nit: let's move this section below `// Terminators for JSON key encoding.` section so that LTREE-specific markers are next to each other.
___
*[`pkg/util/encoding/encoding.go` line 874 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/151597#-OXQWwKK8zOM34RQEAAp:-OXQWwKK8zOM34RQEAAq:b72c7dn) ([raw file](https://github.com/cockroachdb/cockroach/blob/fedea9cd7194c007941f21979b12702c034c5c3a/pkg/util/encoding/encoding.go#L874)):*
> ```Go
> }
> 
> // getBytesLength finds the length of a bytes encoding.
> ```

nit: the comment needs an adjustment.
___
*[`pkg/util/encoding/encoding.go` line 888 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/151597#-OXQX5zy27EEEe-_Q-_S:-OXQX5zy27EEEe-_Q-_T:b-hclqe5) ([raw file](https://github.com/cockroachdb/cockroach/blob/fedea9cd7194c007941f21979b12702c034c5c3a/pkg/util/encoding/encoding.go#L888)):*
> ```Go
> }
> 
> // prettyPrintInvertedIndexKey returns a string representation of the path part of a JSON inverted
> ```

nit: we should add LTREE to `PrettyPrintValueEncoded`. Relatedly, we should include LTREE value type into `TestValueEncodingRand` / `randValueEncode` (maybe also some other tags that are missing there). This should probably be done in a separate PR.
___
*[`pkg/util/encoding/encoding.go` line 889 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/151597#-OXQhDyh06R8wfXz6VRc:-OXQhDyi8OSxDksNHsFM:bchdfjx) ([raw file](https://github.com/cockroachdb/cockroach/blob/fedea9cd7194c007941f21979b12702c034c5c3a/pkg/util/encoding/encoding.go#L889)):*
> ```Go
> 
> // prettyPrintInvertedIndexKey returns a string representation of the path part of a JSON inverted
> // index.
> ```

nit: another comment related to testing: we should adjust `RandDatumWithNullChance` for LTREE to also include digits, hyphens, and underscores.


<!-- Sent from Reviewable.io -->

paulniziolek committed:
sql: add LTREE keyside encoding

This adds basic keyside encoding for LTREE, which is encoded similarly to a
string array: ltrees have their own terminator and their labels are encoded via
escaped bytes.

Informs: cockroachdb#44657
Epic: CRDB-148

Release note: None

paulniziolek reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/151597)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @yuzefovich)

___
*[`pkg/util/encoding/encoding.go` line 888 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/151597#-OXQX5zy27EEEe-_Q-_S:-OXUmCI_Cj27Xh4rzJjz:b-bqk2nu) ([raw file](https://github.com/cockroachdb/cockroach/blob/fedea9cd7194c007941f21979b12702c034c5c3a/pkg/util/encoding/encoding.go#L888)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: we should add LTREE to `PrettyPrintValueEncoded`. Relatedly, we should include LTREE value type into `TestValueEncodingRand` / `randValueEncode` (maybe also some other tags that are missing there). This should probably be done in a separate PR.
</blockquote></details>

Will do. 
___
*[`pkg/util/encoding/encoding.go` line 889 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/151597#-OXQhDyh06R8wfXz6VRc:-OXUmDSaEQ2h7mnLg9Jj:bhy9ot6) ([raw file](https://github.com/cockroachdb/cockroach/blob/fedea9cd7194c007941f21979b12702c034c5c3a/pkg/util/encoding/encoding.go#L889)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: another comment related to testing: we should adjust `RandDatumWithNullChance` for LTREE to also include digits, hyphens, and underscores.
</blockquote></details>

Will do.


<!-- Sent from Reviewable.io -->

paulniziolek reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/151597)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @yuzefovich)



<!-- Sent from Reviewable.io -->

yuzefovich reviewed (approved):
Nice work! <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

Ack on `TestEncDatumCompare` that already does what I was thinking about.

> Reviewed 5 of 6 files at r4, 3 of 3 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/151597#-:-OXVuubK4DOCN4ut0UT1:b-i7g1lx)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @mgartner and @paulniziolek)



<!-- Sent from Reviewable.io -->

paulniziolek commented:
bors r+

mgartner reviewed (commented):
Nice work!

> Reviewed 6 of 9 files at r1, 3 of 6 files at r4, 3 of 3 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/151597#-:-OXZ1Vwi-kJvB3tvah1d:bdtx8r6)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->
